dozer-ingestion/javascript/src/js_extension/mod.rs

```diff
@@ -0,0 +1,164 @@
+use std::{future::Future, sync::Arc};
+
+use deno_runtime::{
+    deno_core::{self, anyhow::Error, extension, op2, ModuleSpecifier},
+    permissions::PermissionsContainer,
+    worker::{MainWorker, WorkerOptions},
+};
+use dozer_ingestion_connector::{
+    dozer_types::{
+        errors::{internal::BoxedError, types::DeserializationError},
+        json_types::serde_json_to_json_value,
+        models::ingestion_types::IngestionMessage,
+        serde::{Deserialize, Serialize},
+        serde_json,
+        thiserror::{self, Error},
+        types::{Field, Operation, Record},
+    },
+    tokio::{runtime::Runtime, task::LocalSet},
+    Ingestor,
+};
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(crate = "dozer_ingestion_connector::dozer_types::serde")]
+pub enum MsgType {
+    SnapshottingStarted,
+    SnapshottingDone,
+    Insert,
+    Delete,
+    Update,
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(crate = "dozer_ingestion_connector::dozer_types::serde")]
+pub struct JsMessage {
+    typ: MsgType,
+    old_val: serde_json::Value,
+    new_val: serde_json::Value,
+}
+
+#[op2(async)]
+fn ingest(
+    #[state] ingestor: &Ingestor,
+    #[serde] val: JsMessage,
+) -> impl Future<Output = Result<(), Error>> {
+    send(ingestor.clone(), val)
+}
+
+extension!(
+    dozer_extension,
+    ops = [ingest],
+    options = { ingestor: Ingestor },
+    state = |state, options| {
+        state.put(options.ingestor);
+    },
+);
+
+pub struct JsExtension {
+    runtime: Arc<Runtime>,
+    ingestor: Ingestor,
+    module_specifier: ModuleSpecifier,
+}
+
+#[derive(Debug, Error)]
+pub enum JsExtensionError {
+    #[error("Failed to canonicalize path {0}: {1}")]
+    CanonicalizePath(String, #[source] std::io::Error),
+}
+
+impl JsExtension {
+    pub fn new(
+        runtime: Arc<Runtime>,
+        ingestor: Ingestor,
+        js_path: String,
+    ) -> Result<Self, JsExtensionError> {
+        let path = std::fs::canonicalize(js_path.clone())
+            .map_err(|e| JsExtensionError::CanonicalizePath(js_path, e))?;
+        let module_specifier =
+            ModuleSpecifier::from_file_path(path).expect("we just canonicalized it");
+        Ok(Self {
+            runtime,
+            ingestor,
+            module_specifier,
+        })
+    }
+
+    pub async fn run(self) -> Result<(), BoxedError> {
+        let runtime = self.runtime.clone();
```

Comment:
Can we use `tokio::task::spawn_blocking()` here to prevent having to pass runtimes around?

Reply:
No, `MainWorker` is `!Send + !Sync`, `worker.execute_main_module` and `worker.run_event_loop` must be `awaited` in a `LocalSet`.

Reply:
Ah, yes, I read over that usage.

All good then